packages/legend-query/src/components/QueryBuilderValueSpecificationEditor.tsx

```diff
@@ -533,75 +587,596 @@ export const LatestDatePrimitiveInstanceValueEditor: React.FC = () => (
+        value: QUERY_BUILDER_DATE_OPTION.PREVIOUS_DAY_OF_WEEK,
+        unit: latestDurationUnitValue,
+      };
+      setValueSpecification(
+        generateDateFunctionExpression(graph, previousDayOfWeekDateOption),
+      );
+      setDateOptionValue({ ...previousDayOfWeekDateOption });
+    }
+  };
+
+  return (
+    <div className="query-builder-value-spec-editor__date-option-dropdown__custom-date">
+      <div className="query-builder-value-spec-editor__date-option-dropdown__custom-date__input">
+        <div className="query-builder-value-spec-editor__date-option-dropdown__custom-date__input-label">
+          Day of Week
+        </div>
+        <CustomSelectorInput
+          className="query-builder-value-spec-editor__date-option-dropdown__custom-date__input-dropdown"
+          options={Object.values(QUERY_BUILDER_DAY_OF_WEEK).map((t) => ({
+            value: t.toString(),
+            label: t.toString(),
+          }))}
+          onChange={(durationUnit: CustomDateInputOption): void => {
+            setDayOfWeekValue(durationUnit.value);
+            updateValueSpecification(durationUnit.value);
+          }}
+          value={{ value: dayOfWeekValue, label: dayOfWeekValue }}
+          darkMode={true}
+        />
+      </div>
+    </div>
+  );
+};
+
 export const DateInstanceValueEditor = observer(
   (props: {
-    valueSpecification: PrimitiveInstanceValue;
+    valueSpecification: PrimitiveInstanceValue | SimpleFunctionExpression;
     graph: PureModel;
-    expectedType: Type;
+    typeCheckOption: TypeCheckOption;
     className?: string | undefined;
+    updateValueSpecification: (val: ValueSpecification | undefined) => void;
   }) => {
-    const { valueSpecification, graph, expectedType, className } = props;
-    const variableType = valueSpecification.genericType.value.rawType;
-    const selectedType = buildElementOption(variableType);
-    const typeOptions: PackageableElementOption<Type>[] = graph.primitiveTypes
-      .filter(
-        (p) =>
-          p.name === PRIMITIVE_TYPE.STRICTDATE ||
-          p.name === PRIMITIVE_TYPE.DATETIME ||
-          p.name === PRIMITIVE_TYPE.LATESTDATE,
-      )
-      .map((p) => buildElementOption(p) as PackageableElementOption<Type>);
-
+    const {
+      valueSpecification,
+      updateValueSpecification,
+      graph,
+      typeCheckOption,
+      className,
+    } = props;
     const strictDate = graph.getPrimitiveType(PRIMITIVE_TYPE.STRICTDATE);
     const date = graph.getPrimitiveType(PRIMITIVE_TYPE.DATE);
     const dateTime = graph.getPrimitiveType(PRIMITIVE_TYPE.DATETIME);
     const latestDate = graph.getPrimitiveType(PRIMITIVE_TYPE.LATESTDATE);
-    const changeType = (val: PackageableElementOption<Type>): void => {
-      if (variableType !== val.value) {
-        genericType_setRawType(valueSpecification.genericType.value, val.value);
-      }
+    const targetQueryBuilderDateOptionsEnum = typeCheckOption.match
```

Comment:
why do we need a specific input of `match` (probably not a good name as i dont know what it is intended for) ? 
Seems like it only depends on the type and whether it is of type `DateTime`.

Reply:
For some cases, we need to filter out some options for DateTime because value of date/strictdate type can't be assigned to DateTime anymore. E.g. the dateTime parameters of derivedProperty. We pass in the value directly without calling isOnDay, isAfterDay.
Comments were added in the file too.